File: elixir/triangle/triangle.exs
defmodule Triangle do
  def kind(a, b, c) do
    cond do
      any_not_positive?(a, b, c) -> {:error, "all side lengths must be positive"}
      is_inequal?(a, b, c) -> {:error, "side lengths violate triangle inequality"}
      is_equilateral?(a, b, c) -> {:ok, :equilateral}
      is_isosceles?(a, b, c) -> {:ok, :isosceles}
      true -> {:ok, :scalene}
    end
  end

  def is_inequal?(a, b, c), do: a + b <= c || a + c <= b || b + c <= a

  def any_not_positive?(a, b, c), do: Enum.any?([a, b, c], fn x -> x <= 0 end)

  def is_isosceles?(a, b, c), do: a == b || a == c || b == c

  def is_equilateral?(a, b, c), do: a == b && b == c
end

File: erlang/triangle/src/triangle.erl
-module(triangle).

-export([kind/3, test_version/0]).

kind(A, B, C) when (A =< 0) or (B =< 0) or (C =< 0) ->
  {error, "all side lengths must be positive"};

kind(A, B, C) when (A + B =< C) or (A + C =< B) or (B + C =< A) ->
  {error, "side lengths violate triangle inequality"};

kind(_A, _A, _A) ->
  equilateral;

kind(_A, _A, _B) ->
  isosceles;

kind(_B, _A, _A) ->
  isosceles;

kind(_A, _B, _A) ->
  isosceles;

kind(_A, _B, _C) ->
  scalene.

test_version() -> 1.
